Mathlib/Algebra/Algebra/TransAlgStruct.lean

```diff
@@ -0,0 +1,159 @@
+/-
+Copyright (c) 2024 Zixun Guo. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Zixun Guo
+-/
+
+import Mathlib.Algebra.Algebra.Defs
+import Mathlib.Algebra.Algebra.Equiv
+
+/-!
+# transfer Algebra structure through ring equiv
+
+In this file, we intro some definitions and lemma to transfer Algebra structure through ring equiv.
+
+## Main Results
+* `ofAlgOnEquivRing`: if `R' ≃+* R`, then `Algebra R' B` can induce a canonical `Algebra R B`.
+* `ofRingEquiv`: if `R' ≃+* R`, then there is a canonical `Algebra R R'` instance.
+* `ofRingEquivReverse`: if `R' ≃+* R`, then there is a canonical `Algebra R' R` instance.
+* `ofAlgOnEquivRing_scalar_tower`: shows that the Algebra structure generated by `ofAlgOnEquivRing`
+    is compatible with the Algebra structure generated by `ofRingEquiv`
+* `of_equiv_on_equiv_ring`: define a `AlgEquiv` directly from equiv ring
+-/
+
+namespace Algebra
+
+/--
+transfer Algebra R' B to Algebra R B if R' ≃+* R
+-/
+noncomputable def ofAlgOnEquivRing
+    {R: Type*} (R' B : Type*)
+    [CommSemiring R] [CommSemiring R'] [Semiring B]
+    [Algebra R' B]
+    (e: R ≃+* R') :
+    Algebra R B where
+smul := fun r b => e r • b
+toFun := fun r => e r • 1
+map_one' := by
+  simp only [map_one, one_smul]
+map_mul' := by
+  intro x y
+  simp only [map_mul, Algebra.mul_smul_comm, mul_one]
+  rw [<-mul_smul, mul_comm]
+map_zero' := by
+  simp only [map_zero, zero_smul]
+map_add' := by
+  intro x y
+  simp only [map_add,add_smul]
+commutes' := by
+  intro r b
+  simp only [RingHom.coe_mk, MonoidHom.coe_mk, OneHom.coe_mk, Algebra.smul_mul_assoc, one_mul,
+    Algebra.mul_smul_comm, mul_one]
+smul_def' := by
+    intro r b
+    simp only [RingHom.coe_mk
+    , MonoidHom.coe_mk, OneHom.coe_mk, Algebra.smul_mul_assoc, one_mul]
+    rfl
```

Comment:
I think the version of this we actually want is
```

variable {R R' A : Type*}

abbrev Algebra.compHom [CommSemiring R] [CommSemiring R'] [Semiring A] [Algebra R' A] (e : R →+* R') :
    Algebra R A where
  smul r b := e r • b
  toRingHom := .comp (algebraMap R' A) e
  commutes' _ _ := Algebra.commutes _ _
  smul_def' _ _ := Algebra.smul_def _ _
```
which matches `Module.compHom` etc.

Reply:
Note this only needs `e` to be a ring homomorphism not an isomorphism.